Feat(#56) blog about caching

Yanich96 opened the pull request:
Blog that discusses caching compilation results for different build systems, so that for the development of effective caching in EO.

Closes: #56

<!-- start pr-codex -->

---

## PR-Codex overview
This PR adds SVG diagrams illustrating different compilation steps and cached files. It includes visual representations of various Mojos in the build process.

### Detailed summary
- Added SVG diagrams for AssembleMojo, SavingInCacheEO, RewritingInCacheEO1, and RewritingInCacheEO2
- Illustrated compilation steps and cached files with visual representations
- Included different Mojos like ParseMojo, OptimizeMojo, ShakeMojo, etc.

> The following files were skipped due to too many changes: `images/RewritingInCacheEO2.svg`, `images/defaultPhaseMaven.svg`, `images/defaultCPhase.svg`, `images/EO.svg`, `_posts/2024/2024-02-06-about-caching-in-eo.md`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

Yanich96 commented:
@maxonfjvipon check please

Yanich96 commented:
@maxonfjvipon check please

Yanich96 commented:
@yegor256 @maxonfjvipon @volodya-lombrozo read please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+
+
+## Introduction
+Wasting a lot of time on building a project is a programming problem. At the moment a programmer starts an
```
@Yanich96 

> Wasting a lot of time on building a project is a programming problem. At the moment a programmer starts an assembly, he loses focus on a task and spends valuable working 

"Empty words". We can remove them without losing any meaning.

volodya-lombrozo replied:
@Yanich96 

> Different build systems use many tools,
 helping to assemble a project faster, namely caching, task parallelization, distributed building and much more.

Why do I need this information?

Yanich96 replied:
@volodya-lombrozo I written it to start this blog. I will delete these suggestions if they are not necessary.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+Wasting a lot of time on building a project is a programming problem. At the moment a programmer starts an
+assembly, he loses focus on a task and spends valuable working time. Different build systems use many tools,
+helping to assemble a project faster, namely caching, task parallelization, distributed building and much more.
+The subject of this article is caching, because completed tasks caching allows not to spend resources again. 
```
@Yanich96 

"The subject of this article is caching."

The other is obvious: 
> because completed tasks caching allows not to spend resources again

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+assembly, he loses focus on a task and spends valuable working time. Different build systems use many tools,
+helping to assemble a project faster, namely caching, task parallelization, distributed building and much more.
+The subject of this article is caching, because completed tasks caching allows not to spend resources again. 
+So in [EO](https://github.com/objectionary/eo) caching is used for speeding up programs work.
```
@Yanich96 Caching speeds up a "build time" or "program execution", not "programs work".

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+helping to assemble a project faster, namely caching, task parallelization, distributed building and much more.
+The subject of this article is caching, because completed tasks caching allows not to spend resources again. 
+So in [EO](https://github.com/objectionary/eo) caching is used for speeding up programs work.
+While developing [EO](https://github.com/objectionary/eo) we found caching errors in `eo-maven-plugin`
```
@Yanich96 Do you have particular links to these issues?

Yanich96 replied:
@volodya-lombrozo do you mean that I should to attach a link to the issue where the error occurred?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+The subject of this article is caching, because completed tasks caching allows not to spend resources again. 
+So in [EO](https://github.com/objectionary/eo) caching is used for speeding up programs work.
+While developing [EO](https://github.com/objectionary/eo) we found caching errors in `eo-maven-plugin`
+for EO version `0.34.0`. The error occurred, because using a file name and comparing equality of
```
@Yanich96 It's hard to grasp without context:

> The error occurred, because using a file name and comparing equality of
 compilation time and caching time is not the most reliable verification.

Yanich96 replied:
@volodya-lombrozo Do you have an example of context? Should it be code or diagram?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+While developing [EO](https://github.com/objectionary/eo) we found caching errors in `eo-maven-plugin`
+for EO version `0.34.0`. The error occurred, because using a file name and comparing equality of
+compilation time and caching time is not the most reliable verification. Unit tests were written showing that 
+cache does not work correctly. Also reading a file was necessary for getting a programme name 
```
@Yanich96 

"Unit tests were written to demonstrate that the cache does not function correctly. Additionally, reading a file was required to obtain a program name, which slowed down the assembly process."

By the way, what is the "assembly proccess"? A reader might not be familiar with this term.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+compilation time and caching time is not the most reliable verification. Unit tests were written showing that 
+cache does not work correctly. Also reading a file was necessary for getting a programme name 
+that slowed down an assembly. 
+That we came to conclusion that we need caching with a reliable verification which does not require reading a file
```
@Yanich96 
1. Came to conclusion" should be "came to the conclusion".
2. "which does not require reading a file from disk" could be rephrased to "that does not require reading a file from a file system".
3. "And using cache" should be "And using a cache"

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+That we came to conclusion that we need caching with a reliable verification which does not require reading a file
+from disk. And using cache should save us enough time for building a project.
+
+The goal of this article is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
```
@Yanich96 This sentence might be connected with the previous one: "The subject of this article is caching."

volodya-lombrozo reviewed (changes requested):
@Yanich96 Please, check all the grammar mistakes in this text.

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+
+
+## Introduction 
+In [EO](https://github.com/objectionary/eo) a caching is used to speed up program execution.
```
@Yanich96 
```txt
In [EO](https://github.com/objectionary/eo), caching is used to speed up program execution.
```

- "caching" is uncountable here, you don't need an article "a"
- You need a comma after "In EO"

volodya-lombrozo replied:
@Yanich96 "program execution" or "program compilation" ?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+
+## Introduction 
+In [EO](https://github.com/objectionary/eo) a caching is used to speed up program execution.
+While developing [EO](https://github.com/objectionary/eo) we found a caching 
```
@Yanich96 No need to mention `EO` twice:
```
While developing [EO](https://github.com/objectionary/eo)
```
In the previous sentence you already mentioned it. Just:
```
Recently we found a error...
```
, btw a "error"? Maybe it's just a "bug"?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+In [EO](https://github.com/objectionary/eo) a caching is used to speed up program execution.
+While developing [EO](https://github.com/objectionary/eo) we found a caching 
+[error](https://github.com/objectionary/eo/issues/2790) in `eo-maven-plugin`
+for EO version `0.34.0`. The error occurred because the cache was searched for the needed file using 
```
@Yanich96 
Please, [use active voice, instead of passive](https://ell.stackexchange.com/questions/87182/active-and-passive-voice):
```
The error occurred because the cache was searched for the needed file using 
 a comparison of compilation time and caching time.
```

> In many styles of writing, active voice is preferred over passive voice for clarity and easier reading.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+a comparison of compilation time and caching time.
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+[Unit tests](https://github.com/objectionary/eo/pull/2749) were written to show that the
```
@Yanich96 
Please, [use active voice, instead of passive](https://ell.stackexchange.com/questions/87182/active-and-passive-voice):
```
The error occurred because the cache was searched for the needed file using 
 a comparison of compilation time and caching time.
```

> In many styles of writing, active voice is preferred over passive voice for clarity and easier reading.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+cache does not work correctly. Additionally, reading a file was necessary to obtain a program name 
+that slowed down the build process. 
+That we came to the conclusion that we need caching with a reliable verification method
+that does not require reading a file system. Using a cache should save us enough time for building a project.
```
@Yanich96 "from a file system"?

volodya-lombrozo replied:
@Yanich96 

> Using a cache should save us enough time for building a project.

We already use some cache. So it already saves some time.
btw, "enough" - How much is that?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,237 @@
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+[Unit tests](https://github.com/objectionary/eo/pull/2749) were written to show that the
+cache does not work correctly. Additionally, reading a file was necessary to obtain a program name 
```
@Yanich96 [Question] To be honest, I don't think we need such details in this post. They only confuse a reader. Maybe it's better to omit them? What do you think?

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+the compilation time and caching time to search for the needed file.
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+That we came to the conclusion that we need caching with a reliable verification method
```
@Yanich96 "That...that" - to many "that`s". You can omit the first one.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+Recently we found a caching 
+[bug](https://github.com/objectionary/eo/issues/2790) in `eo-maven-plugin`
+for EO version `0.34.0`. The error occurred because the algorithm compared
+the compilation time and caching time to search for the needed file.
```
@Yanich96 
This sentence is a bit strange:
> The error occurred because the algorithm compared
 the compilation time and caching time to search for the needed file.

Which algorithm?  Which file do you mean?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+That we came to the conclusion that we need caching with a reliable verification method
+that does not require reading a file system.
```
@Yanich96 Does caching read a file system? Maybe "files from a file system?"

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+that does not require reading a file system.
+
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to create effective caching in [EO](https://github.com/objectionary/eo).
```
@Yanich96 "create" -> "implement"

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to create effective caching in [EO](https://github.com/objectionary/eo).
+
+<!--more-->
```
@Yanich96 "More"?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+## Build caching of existing build systems
+
+### ccache/sccache
+In compiled programming languages, building a project takes a long time.
```
@Yanich96 "long time" ? How much is it? I build all my projects relatively fast.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+
+## Build caching of existing build systems
+
+### ccache/sccache
```
@Yanich96 Is it a build system or what? Where is the link? Short description?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+  <img src="/images/ccache.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are code files and header files.
```
@Yanich96 I didn't find "preprocessor" on the picture above.
Moreover, I though that "header files" are code files too.

Yanich96 replied:
![preprocessor](https://github.com/objectionary/news.eolang.org/pull/58/files#diff-e33835f974f6fe763616c951e4b40a95224275dea98014f386a38fb2c0cb88d2)

Yanich96 replied:
@volodya-lombrozo "preprocessor" on the diagram. I didn't got you

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+
+1) First, preprocessor gets the input files. The input files are code files and header files.
+The preprocessor removes comments from the code and converts the code in accordance
+with macros and executes other directives, starting with the “#” symbol
```
@Yanich96 No need to describe how exactly a preprocessor works. It's important that we get at the end of this phase.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+
+2) The compiler receives the finished code file and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language. Next, the parsing occurs preprocessor code into machine code
```
@Yanich96 What does it mean?

>  Next, the parsing occurs preprocessor code into machine code according to the rules.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language. Next, the parsing occurs preprocessor code into machine code
+according to the rules.
+At the end of its work, the compiler optimizes the resulting machine code and produces an object file. 
```
@Yanich96 "At the end of its work" -> "At the end"

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+according to the rules.
+At the end of its work, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel,
+that is, we receive several object files at once.
```
@Yanich96 This is redundant:

>  that is, we receive several object files at once

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+To speed up compilation, different files of the same project are compiled in parallel,
+that is, we receive several object files at once.
+
+3)  After all received project object files are passed to the linker.
```
@Yanich96 What does it mean:

> After all received project object files are passed to the linker.

Is it "After all, received project object files are passed to the linker."
or "After, all received project object files are passed to the linker." ?
Why "received"? Which "project" do you mean? 

Maybe it's better just use "Then, object files are passed to the linker.", or better:
"Then linker <...do something...> with object files" (active voice)?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+that is, we receive several object files at once.
+
+3)  After all received project object files are passed to the linker.
+Linker is a program that combines program components, written in assembly language or a high-level programming language,
```
@Yanich96 I though that Linker combines object files?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+
+
+`ccache` has two main caching methods:
+1) `Direct mode` - hashcode is generated based on the source code.
```
@Yanich96 Which "hashcode" do you mean? You gave the definition below, that paragraph positioning confuses a lot. I have to skip this part and then return to it after.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,235 @@
+The hashcode includes information: file contents, directory, compiler information, compilation time, extensions
+used by the compiler. A compressed machine code file is placed in the cache using the received key.
+
+`Direct mode` compiles the program faster, since the preprocessor step is skipped. 
```
@Yanich96 You explains two modes by using this template:

```
1) Direct mode - hashcode is generated based on the source code.
2) Preprocessor mode - hashcode is generated based on the result of preprocessor.
3) Direct mode compiles the program faster...
4) Preprocessor mode is slower...
```

Looks strange, maybe it's better to explain one mode and the move to the another?

```
1) Direct mode - hashcode is generated based on the source code.
3) Direct mode compiles the program faster...
2) Preprocessor mode - hashcode is generated based on the result of preprocessor.
4) Preprocessor mode is slower..
```

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,239 @@
+
+### Maven
+[Maven](https://maven.apache.org) automates and manages Java-project builds. 
+Building a project in `Maven` is completed in three
```
@Yanich96 in three maven LifeCycles Maven. "Maven, maven"

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,239 @@
+  <img src="/images/defaultPhaseMaven.svg">
+</p>
+
+In `Maven` all phases and goals are executed strictly in order, linearly.
```
@Yanich96 So, Maven doesn't use caching at all?

Yanich96 replied:
@volodya-lombrozo As far as I understand, that Maven can use added extensions from Gradle for caching. Or Maven can rebuild only changed project modules.

volodya-lombrozo replied:
@Yanich96 Maven has `.m2` folder at least. In this folder it keeps all downloaded dependencies. So it's some sort of caching too.

volodya-lombrozo replied:
@Yanich96 It's not about Gradle, I guess: https://maven.apache.org/extensions/maven-build-cache-extension/

volodya-lombrozo reviewed (changes requested):
@Yanich96 Could you please link this PR with the issue you are trying to solve?

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,239 @@
+This machine code is then combined into one executable file.
+
+
+`ccache` uses hashcode to find cached files. The hashcode includes information:
```
@volodya-lombrozo Thanks, I will use "hash algorithm" as [ccache documentation](https://ccache.dev/manual/4.8.2.html#_common_hashed_information).

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,239 @@
+`ccache` has two main caching methods:
+1) `Direct mode` - hashcode is generated based on the source code. 
+`Direct mode` compiles the program faster, since the preprocessor step is skipped.
+However,the header files are not checked for changes, so the wrong project may be built.
```
@volodya-lombrozo 
"the wrong project" - is the project built with not verified header files.
"the right project" - is the project built with verified header files.

volodya-lombrozo replied:
@Yanich96 Maybe we can clarify it in the text?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+  <img src="/images/defaultCPhase.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are source files `.cpp` and header files `.h`.
```
@Yanich96 Wdyt? 
"First, the preprocessor retrieves the source code files, which consist of both source files `.cpp` and header files `.h`."

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+</p>
+
+1) First, preprocessor gets the input files. The input files are source files `.cpp` and header files `.h`.
+The result is a single edited file `.cpp` with human-readable code that the compiler will get.
```
@Yanich96 "edited"? Seems redundant.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+The result is a single edited file `.cpp` with human-readable code that the compiler will get.
+
+
+2) The compiler receives the edited code file `.cpp` and converts it into machine code, presented in an object file.
```
@Yanich96 Is an "object file" machine code?

Yanich96 replied:
@volodya-lombrozo "object file" is machine code.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+
+
+2) The compiler receives the edited code file `.cpp` and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
```
@Yanich96 Wdyt about "[parsing](https://en.wikipedia.org/wiki/Parsing) checks ..."

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel.
```
@Yanich96 I guess it's better to say "might be compiled".

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example
```
@Yanich96 To be honest, I have some doubts about this paragraph where you discuss "compilation steps":
1) First of all this is a blog about caching, not about compilation
2) You describe compilation incompletely. What about optimizations? Moreover, modern compilers usually convert source code to some sort of IR, like LLVM IR, for example. You can take a look how [clang](https://en.wikipedia.org/wiki/Parsing) works.

Yanich96 replied:
@volodya-lombrozo I believe that a reader will better understand this article if we briefly talk about the stages of compilation of the presented build systems. Yes, I describe compilation incompletely, but enough to indicate where caching works.

volodya-lombrozo replied:
@Yanich96 Then, it's good to mention it.

Yanich96 replied:
@volodya-lombrozo "The goal is to implement effective caching in [EO](https://github.com/objectionary/eo).
For this, we will briefly look at how frequently used build systems (`ccache`, `Maven`, `Gradle`) work
in order to better understand the ideas behind caching in them."
 it's ok?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+This machine code is then combined into one executable file.
+
+
+`ccache` hash algorithm, for the hashing of information to find cached files fast. 
```
@Yanich96 This sentence is inconsistent with the previous one. Moreover, it seems we have grammar errors here.

Yanich96 replied:
@volodya-lombrozo the two last sentences (lines 52-55) was unnecessary. I need delete them. Thanks, I have found grammar error.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+1) `Direct mode` - hash is generated based on the source code. 
+`Direct mode` compiles the program faster, since the preprocessor step is skipped.
+However,the header files are not checked for changes, so the project may be built with not verified header files.
+2) `Preprocessor mode` - hash is generated based on the result of preprocessor.
```
@Yanich96 Is it some particular result ("the")? Which preprocessor do you mean?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in a cloud data storage.
+And `sccache` includes support for caching the compilation of C/C++ code, 
+[Rust](https://github.com/mozilla/sccache/blob/main/docs/Rust.md), as well as NVIDIA's CUDA using 
+[nvcc](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html),
```
@Yanich96 What is `nvcc` you didn't mentioned it before.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,263 @@
+And `sccache` includes support for caching the compilation of C/C++ code, 
+[Rust](https://github.com/mozilla/sccache/blob/main/docs/Rust.md), as well as NVIDIA's CUDA using 
+[nvcc](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html),
+and [clang](https://llvm.org/docs/CompileCudaWithLLVM.html), while `ccache` works with C and C++ code.
```
@Yanich96 What do you mean:

> `ccache` works with C and C++ code.

As I undestand `ccache` works with compilers, not with code.

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,251 @@
+The result of the linker is an executable `.exe` file.
+
+    
+`ccache` has hash algorithm, for the hashing of information to find cached files fast. 
```
@Yanich96 This paragraph is disconnected with the previous one. You only said about compilation and then suddenly started with `ccache`.  Please, read this two paragraphs and you will see.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,251 @@
+### Maven
+[Maven](https://maven.apache.org) automates and manages Java-project builds. 
+Building a project in `Maven` is completed in three
+[Maven LifeCycles](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html),
```
@Yanich96 Are you sure here? Yes, Maven definitely has three life cycles. However, to build a project you might use the only one.

Moreover, "Building a project in `Maven` is completed" has some particular [meanings](https://dictionary.cambridge.org/dictionary/english/complete). None of them are suitable here.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,251 @@
+`Maven` suggests rebuilding only changed project modules to speed up the build process.
+
+### Gradle
+But unlike `Maven`, [Gradle](https://gradle.org) builds projects using a task graph - 
```
@Yanich96 Just "Unlike `Maven`..."

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,251 @@
+
+### Gradle
+But unlike `Maven`, [Gradle](https://gradle.org) builds projects using a task graph - 
+[Directed Acyclic Graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph),
```
@Yanich96 Maybe it's better to give a link to a "Gradle task graph" instead? Why do I need to read about DAGs?

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+1) Before executing a task for the first time, `Gradle` takes a 
+   [fingerprint](https://en.wikipedia.org/wiki/Fingerprint_(computing))
+   of the path and contents of the source files and saves it.
+2) Then `Gradle` executes the task and saves a fingerprint of the path and contents of the output files.
```
@Yanich96

1. **the** fingerprint
2. "of the path and contents" - better to remove it, you already mentioned it above.
3.  Where does it save the fingerprint?

Yanich96 replied:
"Where does it save the fingerprint?" - I don't know. [the Gradle documentation](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work) has the sentence: "Gradle persists both fingerprints for the next time the task is executed." Nothing is said about the save location.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+   [fingerprint](https://en.wikipedia.org/wiki/Fingerprint_(computing))
+   of the path and contents of the source files and saves it.
+2) Then `Gradle` executes the task and saves a fingerprint of the path and contents of the output files.
+3) Before each rebuilding of the task, `Gradle` generates a new fingerprint of the source files
```
@Yanich96 "Then, when `Gradle` starts a project build again, it generates a new fingerprint for the same files. If the new fingerprint has not changed, `Gradle` can safely skip this task."

Yanich96 replied:
@volodya-lombrozo "Then, when Gradle starts a project build again, it generates a new fingerprint for the same files.
   If the new fingerprint has not changed, Gradle can safely skip this task can reuse outputs. 
   In the opposite case, the task needs to perform an action and to rewrite outputs.
   The fingerprint is considered current if the last modification time 
   and the size of the source files have not changed." - is it ok?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+   If none of the inputs or outputs have changed, Gradle can skip that task.
+
+
+In addition to `Incremental build`, `Gradle` also stores fingerprints of previous builds, enabling quick project builds,
```
@Yanich96 What is the difference with the cache that you described above?

Yanich96 replied:
@volodya-lombrozo this cache  is for various branches, above - for project of one branch

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+
+Moreover, `ccache` has two types of the hashing:
+1) `Direct mode` - the hash is generated based on the source code only. 
+This mode allows to build the program faster, since the preprocessor step is skipped.
```
@volodya-lombrozo if the source code has not been changed, then the result of the preprocessor is taken from the cache

Yanich96 replied:
@volodya-lombrozo I have fixed the text so: "
When compiling a file, its hash is calculated. 
If the file is already present in the registry of compiled files, the file will not be compiled again.
Instead, the previously compiled binary file will be utilized.
This approach can significantly accelerate the build process of certain packages, reducing build times by 5-10 times.
The [`ccache` hash](https://ccache.dev/manual/4.8.2.html#_common_hashed_information) is
based on:
* the file contents
* the current directory of the file
* the name of the compiler
* the compiler’s size and modification time
* extensions used by the compiler.

Moreover, `ccache` has two types of the hashing:
1) `Direct mode` - the hash is generated based on the source code only.
When using this mode, the user must be sure that the external libraries, using in a project, have not changed.
Otherwise, the project will build with errors.
2) `Preprocessor mode` - hash is generated based on the `.cpp` file received after the preprocessor step.
`Preprocessor mode` is slower than `direct mode`, but the project is built without errors."

Yanich96 replied:
@volodya-lombrozo I added an explanation of what `ccache` does: "When compiling a file, its hash is calculated.
If the file is already present in the registry of compiled files, the file will not be compiled again.
Instead, the previously compiled binary file will be utilized.
This approach can significantly accelerate the build process of certain packages, reducing build times by 5-10 times." And I changed the sentences about modes.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+When using this mode, the user must be sure that the external libraries, using in a project, have not changed.
+Otherwise, the project will build with errors.
+2) `Preprocessor mode` - hash is generated based on the `.cpp` file received after the preprocessor step.
+`Preprocessor mode` is slower than `direct mode`, but the project is built without errors.
```
@volodya-lombrozo It is ok: "`Preprocessor mode` is slower than `direct mode`, but the project is built without compile errors."?

Yanich96 replied:
@volodya-lombrozo I will delete these sentence.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
````diff
@@ -0,0 +1,285 @@
+showcasing how inputs and outputs are specified to enable `Incremental build`:
+```
+task myTask {
+    inputs.dir 'src/main/java/MyTask.somebody' // Specify the input directory
````
@volodya-lombrozo I have fixed this example:
```
task myTask {
    inputs.file 'src/main/java/MyTask.somebody' // Specify the input file
    outputs.file 'build/classes/java/main/MyTask.somebody' // Specify the output file
    
    doLast {
        // Task actions go here
        // This code will only be executed if the inputs or outputs have changed
    }
}
```

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
````diff
@@ -0,0 +1,285 @@
+    }
+}
+```
+`Gradle` uses this information to determine if a task is up-to-date and needs to perform any work.
````
@volodya-lombrozo I think that I will delete this sentence

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+`Gradle` uses this information to determine if a task is up-to-date and needs to perform any work.
+
+To understand how `Incremental build` works, consider the following steps:
+1) Before executing a task for the first time, `Gradle` takes a 
```
@volodya-lombrozo I used "take" because this verb is used [the Gradle documentation](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work)

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in a cloud data storage.
+And `sccache` supports a wider range of languages, while `ccache` focuses on caching C and C++ compiler.
+
```
@volodya-lombrozo The principle of caching in `sccache`  is the same as in `ccache` (using Direct and Preprocessor modes), the only difference is in the places where the data is stored.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+
+### Conclusion
+In this article, we explored various build systems and their caching methods.
+We were motivated to find an efficient caching approach for EO due to issues discovered during bug investigation.
```
@Yanich96 To be honest, I don't understand why you "explored various build systems". This analysis is completely disconnected from the solution you propose. You observe high-level concepts in other build systems but fail to draw any conclusions from them. You merely mention their existence. Why? Why should we read about them? Where is the connection with the second part of the text?

Later, you just mentioned that we have some goals and some Mojos. Why? Why do we need to read about them? What is the purpose of this?

Furthermore, you only mention some classes. Why do they have this exact architecture? Why delve into such low-level details here? Perhaps it's better to suggest a high-level solution instead?

Please, pause for a moment and consider these questions:

What is the purpose of this blog post? Why are you writing it, and why should people read it?
At which level of abstraction do you want to discuss? High-level caching mechanisms like ccache, or low-level CacheValidation implementation?

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+`Gradle` employs
+[Incremental build](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work),
+to speed up project builds.
+For an incremental build to work, the tasks used to build the project must have specified
```
@volodya-lombrozo "The tasks that build the project must have input and output files for an incremental build to work." - is it ok?

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+* `the cached file`: This file contains the results of goal's execution.
+
+
+The previous caching mechanism in EO made use of distinct interfaces, specifically `Footprint` and `Optimization`,
```
@volodya-lombrozo It's my mistake. There was `SafeMojo` in another sentence.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,285 @@
+
+
+In this chapter, we introduce the keywords:
+* `the source file`: This file serves as the input for goal operations.
```
@volodya-lombrozo I will remove these text.

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+3) Then, when Gradle starts a project build again, it generates a new fingerprint for the same files.
+   If the new fingerprint has not changed, Gradle can safely skip this task. 
+   In the opposite case, the task needs to perform an action and to rewrite outputs.
+   The fingerprint is considered current if the last modification time 
```
@Yanich96 This sentence makes no sense to me. If you have an intention to describe "fingerprint". It's better to do above where you left the link to the fingerprint definition. You might say that "fingerprint is based on last modification time and size of the source files"

Yanich96 replied:
@volodya-lombrozo I did this:
1) Before executing a task, `Gradle` takes a 
   [fingerprint](https://en.wikipedia.org/wiki/Fingerprint_(computing))
   of the path and contents of the inputs files and saves it. 
   The fingerprint is considered current if the last modification time
   and the size of the source files have not changed.
2) Then `Gradle` executes the task and saves a fingerprint of the path and contents of the output files.
3) Then, when Gradle starts a project build again, it generates a new fingerprint for the same files.
  If the new fingerprint is current, Gradle can safely skip this task. 
   In the opposite case, the task performs an action again and rewrites outputs.

I believe that the sentence  "fingerprint is based on last modification time and size of the source files" is not good idea because fingerprint is based on the path and the contents of  a file.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+
+
+In addition to `Incremental build`, `Gradle` also stores fingerprints of previous builds, enabling quick project builds,
+for example when switching from one branch to another. This feature is known as 
```
@Yanich96 Please, specify which branch do you mean. Is it a "git" branch?

Yanich96 replied:
@volodya-lombrozo yes, "git" branch

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+Each `module` has its own `pom.xm` file, and there is an aggregator `pom.xml` that consolidates all the `modules`.
+This plugin takes a key for a `module`, it encapsulates the essential aspects of the `module`,
+including the source code and the configuration of the plugins used within it.
+`Modules` with the same key are current or unchanged and the cache can efficiently restore them.
```
@Yanich96 This sentence doesn't make sense to me: "`Modules` with the same key are current or unchanged and the cache can efficiently restore them.".
By the way, what is the "key"?

Yanich96 replied:
@volodya-lombrozo key is hash. Can you explain me your point: "This sentence doesn't make sense to me"? Do you mean this is obvious?

volodya-lombrozo replied:
@Yanich96  I don't understand "current" word here

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+This plugin takes a key for a `module`, it encapsulates the essential aspects of the `module`,
+including the source code and the configuration of the plugins used within it.
+`Modules` with the same key are current or unchanged and the cache can efficiently restore them.
+Conversely, the cache seamlessly delegates the build work to the standard Maven core,
```
@Yanich96 "Conversely"?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+These caching interfaces shared similar logic, but with minor differences.
+For instance, `Footprint` verifies the EO version of the compiler, whereas the remaining checks are identical.
+Additionally, the conditions for searching data in the cache had errors. 
+The cached file is considered valid if the end time of goal's execution
```
@Yanich96 You don't need to explain it again. I've already mentioned this bug above. Could you please remove this sentence?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+from the file attributes without reading the file context.
+
+
+### Conclusion
```
@Yanich96 I don't think we need such a conclusion in a blog post. It isn't a scientific article. Moreover it doesn't provide any useful information. Kinda "water".

volodya-lombrozo reviewed (changes requested):
@Yanich96 

I would suggest to add conclusion after each build system you mentioned:

1. for ccache
2. for grade
3. for maven

Maybe you will need to move some conclusions from "EO build cache".

"EO build cache" should be a conclusion for the entire blog post where you describe how we will implement caching (on a high-level) in EO.
____

P. S. I understand that you are trying to use different terms from different systems:
- fingerprint
- hash
- key
But sometimes it's hard to keep in mind that they mean the same. So it confuses a bit.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+
+<!--more-->
+
+## Build caching of existing build systems
```
@volodya-lombrozo I will choose " Caching in Other Build Systems"

Yanich96 commented:
@volodya-lombrozo I have fixed this post. 
*I added "between goals" in the sentence in introduction "Recently we found a caching [bug](https://github.com/objectionary/eo/issues/2790) between goals in `eo-maven-plugin` for EO version `0.34.0`."  to make it clear at what level the EO code works.
*I have added conclusions after each build system.
Check please.

volodya-lombrozo reviewed (changes requested):
@Yanich96, please thoroughly review the **entire** text at least once (preferably more) before submitting it for review. This will greatly expedite our review process.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,233 @@
+
+
+`ccache` is a high-level tool and cannot work with individual compilation tasks,
+therefore `ccache` is not suitable for solving our problems.
```
@volodya-lombrozo 
"`ccache` is a high-level tool and cannot work with individual compilation tasks (e.g. `Maven goal` or `Gradle task`).
However, the hashing approach and the concept of non-local data storage could potentially
be incorporated during the development of the EO caching mechanism."
is it ok?

volodya-lombrozo replied:
@Yanich96 Why do you need to mention this:
> is a high-level tool and

?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,233 @@
+
+To understand how `Incremental build` works, consider the following steps:
+1) Before executing a task, `Gradle` takes a hash of the path and contents of the inputs files and saves it.
+   The hash is considered current if the last modification time
```
@Yanich96 I'm not sure what you mean here, but I guess, yes.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,233 @@
+1) Before executing a task, `Gradle` takes a hash of the path and contents of the inputs files and saves it.
+   The hash is considered current if the last modification time
+   and the size of the source files have not changed.
+2) Then `Gradle` executes the task and saves a hash of the path and contents of the output files.
```
@volodya-lombrozo yes. It is written in [The Gradle documentation](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work) - "Gradle takes a fingerprint of the inputs. This fingerprint contains the paths of input files and a hash of the contents of each file. Gradle then executes the task."

volodya-lombrozo replied:
@Yanich96 It would be nice

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,233 @@
+the [Build Cache](https://docs.gradle.org/current/userguide/build_cache.html).
+
+
+The concept of `Gradle Incremental build` bears resemblance to a tool that is essential for our purposes.
```
@volodya-lombrozo The conclusion of `Gradle` section : 
`Gradle Incremental build` can manage separate compilation tasks based on inputs and outputs.
And the EO compiler consists from a unit of work in `Maven` (the last section contains a detailed description).
Steps of the EO compiler can have input and output files.
Building upon the concept of `Gradle Incremental Build`, we can use its principles to develop the EO caching mechanism.

it's all right?

volodya-lombrozo replied:
@Yanich96 it's a bit better, but you can clarify it, I believe.

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+The result is a single file `.cpp` with human-readable code that the compiler will get.
+2) The compiler receives the file `.cpp` from the preprocessor and compiles it into an object file - `.obj`.
+At the compilation stage, parsing checks whether the code matches rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
```
@Yanich96 You already mentioned it:

> The compiler receives the file `.cpp` from the preprocessor and compiles it into an object file - `.obj`.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+[Incremental build](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work),
+to speed up project builds.
+For an incremental build to work, the tasks used to build the project must have specified
+input and output files.
```
@Yanich96 "To enable an incremental build, the tasks that build the project must specify their input and output files."

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
````diff
@@ -0,0 +1,283 @@
+```
+
+
+To understand how `Incremental build` works, consider the following steps:
````
@Yanich96 Something strange is happening here with punctuation. Did you put this sentences in this order intentionally?

Yanich96 replied:
@volodya-lombrozo If I replace "To understand how `Incremental build` works, consider the following steps" with "How `Incremental build` works", will it be ok?

volodya-lombrozo replied:
@Yanich96 Is it possible to just remove this sentence?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+
+To understand how `Incremental build` works, consider the following steps:
+`Incremental build` uses a hash to detect changes in the inputs and the outputs.
+The single hash contains the paths and the contents of all the input files or output files.
```
@Yanich96 "contains"? Maybe "uses"?

Yanich96 replied:
@volodya-lombrozo In [Gragle documentation](https://docs.gradle.org/current/userguide/incremental_build.html#sec:how_does_it_work): "This fingerprint contains the paths of input files and a hash of the contents of each file."

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+
+`Gradle Incremental build` can manage separate compilation tasks based on inputs and outputs.
+And the EO compiler consists from a unit of work in `Maven` (the last section contains a detailed description).
+Steps of the EO compiler can have input and output files.
```
@Yanich96 Why did you write this two sentences about eo?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+
+In Maven, the `phases` are inherently interconnected within the build lifecycle.
+A `phase` represents a specific task, and the execution order of `phases` is determined by the default Maven 
+lifecycle bindings. Each `phase` functions as a series of individual tasks known as `goals`.
```
@Yanich96 You have already described `phases` and `goals` above. Could you please remove this redundancy and repetition?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+It does not use hashing algorithms, which can slow down project build times,
+and it does not have separate cache directories.
+Each `builder` has own directories for input and output data related to their work.
+The operational principle of the Takari Incremental API is similar to the operation of caching in EO.
```
@Yanich96 We don't know about caching in EO yet.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+which consist of both source files `.cpp` and header files `.h`.
+The result is a single file `.cpp` with human-readable code that the compiler will get.
+2) The compiler receives the file `.cpp` from the preprocessor and compiles it into an object file - `.obj`.
+At the compilation stage, parsing checks whether the code matches rules of a specific programming language.
```
@volodya-lombrozo yes, thanks

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+</p>
+
+
+<p align="center">
```
@Yanich96, I agree that we need to remove the redundancy in the code. For example, we should combine the `Footprint` and `Optimization` methods and fix the hash comparison mechanism. As for checking the "previous" step, I completely disagree. First of all, some steps might be skipped, and more importantly, doing this significantly increases [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) between the phases, which is a significant architectural flaw.

volodya-lombrozo replied:
@Yanich96 [This](https://github.com/objectionary/news.eolang.org/pull/58/files#r1602623769) comment is still relevant.

Yanich96 commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,283 @@
+
+Special attention should be given to the Takari Incremental API. 
+This API can be applied to cache EO compilation stages as it operates with `goals`.
+It does not use hashing algorithms, which can slow down project build times,
```
@volodya-lombrozo  The Takari checks the last modification time of the input files. It doesn't create a hash. 

Or did I not understand the question?

volodya-lombrozo replied:
@Yanich96 It would be good to mention it here:
> The Takari checks the last modification time of the input files. It doesn't create a hash.

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,261 @@
+It's also possible to add a new `goal` to a desired phase by modifying the `pom.xml` file. 
+Additionally, Maven also supports `goals` that are not bound to any build phase
+and can be executed outside the build lifecycle, directly through the command line.
+
```
@Yanich96 Do you need this empty line?

Yanich96 replied:
@volodya-lombrozo I guess so, because I talk since in lines 122-128 about Maven and in lines 131-150 about caching mechanisms in Maven.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,261 @@
+which introduces the concept of `builders`. These `builders` are user-provided public non-abstract
+top-level classes that implement specific build actions.
+They can produce various types of outputs, including generated/output files on the filesystem, 
+build messages, and project model mutations. For each `builder` annotated method, a maven mojo, 
```
@Yanich96 I didn't understand this sentence:
```
For each `builder` annotated method, a maven mojo, which represents a maven `goal`, is generated.
```

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,261 @@
+which represents a maven `goal`, is generated.
+When a `builder` is run for a given set of inputs, it produces and saves to the specified directory the same outputs. 
+Any changes in the inputs result in the removal of outputs.
+
```
@Yanich96 Do you need this empty line?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,233 @@
+These tasks happen one after the other, and each task relies on the output of the one before it.
+Each task has directories for input and output data, as well as a directory for storing cached data.
+Using the program name, each task can receive and store data.
+
```
@Yanich96 Same question.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,261 @@
+1) We create EO program, named "example".
+   Intermediate files during compilation of this program will have the same name, but not the format
+   (e.g. `example.eo`, `example.xml`).
+2) When the EO compiler compiles this program task, it saves files of compilation steps into cache.
```
@Yanich96 "compiler compiles"

Yanich96 replied:
@volodya-lombrozo "compiler assembles" is it ok?

Yanich96 commented:
@volodya-lombrozo check please

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+* Employing multiple caching mechanisms for similar entities is a suboptimal practice, 
+leading to redundancy and complicating the caching infrastructure.
+
+In tackling caching challenges within EO, we conducted a thorough evaluation of current caching systems.
```
@Yanich96 There is lot of "water" in this paragraph. Could you simplify this text please?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+for storing and retrieving data from the cache.
+The logic for checking the relevance of cached data is presented below:
+1) We create EO program, named "example".
+   Intermediate files during compilation of this program will have the same name, but not the format
```
@Yanich96 Why do we need to have different format?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+   When the EO compiler assembles this program task, it saves files of compilation steps into cache.
+   Each compilation step has its own caching directory and an input file directory.
+2) When the EO compiler starts a project build again, it will check if there is the input file, named "example",
+   in the cache of step. If such a file exists,
```
@Yanich96 "cache of step"?

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+   Each compilation step has its own caching directory and an input file directory.
+2) When the EO compiler starts a project build again, it will check if there is the input file, named "example",
+   in the cache of step. If such a file exists,
+   then it is enough to check that the last modification time of cached file at the current step
```
@Yanich96 Which files do you compare? According with this text, you compare input with output. Their modification times will be definitely different.

volodya-lombrozo commented on _posts/2024/2024-02-06-about-caching-in-eo.md:
```diff
@@ -0,0 +1,241 @@
+4) If the EO program file [Picture 5](/images/RewritingInCacheEO1.svg) 
+   or any input file [Picture 6](/images/RewritingInCacheEO2.svg) have changed,
+   then the previously cached files become invalid.
+   In this case, the compilation step performs an action again and rewrites outputs.
```
@Yanich96 "rewrite outputs"? Do you mean cache here?

volodya-lombrozo commented:
@yegor256 Could you take a look, please?

yegor256 commented:
@Yanich96 I appreciate the study you conducted with existing caching tools! Great work! In order to make this blog post more impressive and informative I would suggest slightly modify its structure. How about this one:

* What "build caching" is for?
* What existing approaches to build caching we know? (illustrate each approach with an example and a link to a system that uses it)
* What approach did we choose for EO and why?
* How to manage build caching in EO? (show examples of turning it off, checking its status, etc.)
* What are the known limitations (possible downsides) of our solution?
* How much performance gain build caching gives us in EO projects?

WDYT? @volodya-lombrozo

volodya-lombrozo commented:
> @Yanich96 I appreciate the study you conducted with existing caching tools! Great work! In order to make this blog post more impressive and informative I would suggest slightly modify its structure. How about this one:
> 
> * What "build caching" is for?
> * What existing approaches to build caching we know? (illustrate each approach with an example and a link to a system that uses it)
> * What approach did we choose for EO and why?
> * How to manage build caching in EO? (show examples of turning it off, checking its status, etc.)
> * What are the known limitations (possible downsides) of our solution?
> * How much performance gain build caching gives us in EO projects?
> 
> WDYT? @volodya-lombrozo

@yegor256 I don't think we need to stick to the scope of this article without adding something new. It's already large. So, I would exclude this points:

> * What are the known limitations (possible downsides) of our solution?
> * How much performance gain build caching gives us in EO projects?

With all the rest I completely agree. 

@Yanich96 @yegor256  I would also recommend to move this changes into a separate PR since this one takes too much time to load on my laptop. (Because there are many comments already, I guess.)
